control/controlclient/noise.go

```diff
@@ -0,0 +1,146 @@
+
+package controlclient
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/net/http2"
+	"tailscale.com/control/controlbase"
+	"tailscale.com/control/controlhttp"
+	"tailscale.com/types/key"
+	"tailscale.com/util/multierr"
+)
+
+// noiseConn is a wrapper around controlbase.Conn.
+// It allows attaching an ID to a connection to allow
+// cleaning up references in the pool when the connection
+// is closed.
+type noiseConn struct {
+	*controlbase.Conn
+	id   int
+	pool *noiseClient
+}
+
+func (c *noiseConn) Close() error {
+	if err := c.Conn.Close(); err != nil {
+		return err
+	}
+	c.pool.connClosed(c.id)
+	return nil
+}
+
+// noiseClient provides a http.Client to connect to tailcontrol over
+// the ts2021 protocol.
+type noiseClient struct {
+	*http.Client // HTTP client used to talk to tailcontrol
+	privKey      key.MachinePrivate
+	serverPubKey key.MachinePublic
+	connCount    int
+
+	serverHost string // the host:port part of serverURL
+
+	mu       sync.Mutex
+	connPool map[int]*noiseConn // All created connections. TODO(maisem): clean up dead connections.
+}
+
+// newNoiseClient returns a new noiseClient for the provided server and machine key.
+// serverURL is of the form https://<host>:<port> (no trailing slash).
+func newNoiseClient(ctx context.Context, priKey key.MachinePrivate, serverPubKey key.MachinePublic, serverURL string) (*noiseClient, error) {
+	u, err := url.Parse(serverURL)
+	if err != nil {
+		return nil, err
+	}
+	var host string
+	if u.Port() != "" {
+		// If there is an explicit port specified use it.
+		host = u.Host
+	} else {
+		// Otherwise, controlhttp.Dial expects an http endpoint.
+		host = fmt.Sprintf("%v:80", u.Hostname())
+	}
+	np := &noiseClient{
+		serverPubKey: serverPubKey,
+		privKey:      priKey,
+		serverHost:   host,
+	}
+
+	// Create a new http.Client which dials out using nc.Dial.
+	np.Client = &http.Client{
+		Transport: &http2.Transport{
+			DialTLS:   np.Dial,
+			AllowHTTP: true,
+		},
+	}
+
+	return np, nil
+}
+
+// connClosed removes the connection with the provided ID from the pool
+// of active connections.
+func (nc *noiseClient) connClosed(id int) {
+	nc.mu.Lock()
+	defer nc.mu.Unlock()
+	delete(nc.connPool, id)
+}
+
+// Close closes all the underlying noise connections.
+// It is a no-op and returns nil if the connection is already closed.
+func (nc *noiseClient) Close() error {
+	nc.mu.Lock()
+	conns := nc.connPool
+	nc.connPool = nil
+	nc.mu.Unlock()
+
+	var errors []error
+	for _, c := range conns {
+		if err := c.Close(); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return multierr.New(errors...)
+}
+
+// Dial opens a new connection to tailcontrol, fetching the server noise key
+// if not cached. The network and address params are ignored as it only dials
+// the server it was created for.
+func (nc *noiseClient) Dial(_, _ string, cfg *tls.Config) (net.Conn, error) {
```

Comment:
Make this function unexported? I was very confused about it until I realized that it's an internal implementation detail of the http2 Transport configuration.

Also still confused by why we need all this machinery to manage multiple connections, when presumably we're dialing a single destination and (I thought) http2 would provide the multiplexing on top of the Conn.

I'm worried about this specifically because it's going to break some things we could do with noise connections, like binding hardware attestations to the underlying Noise state. We can't do that if we might have 15 different conns kicking around, my assumption was that we'd be running http2 muxing over a single conn, so it would be tractable to bind cryptographic state to the underlying connection.

Reply:
IOW: with noise connections, each Conn is stateful and can carry more contextual information than just what was initially handshaken. Stuff like "client has proven to me, _on this specific Conn_, that they're running on the expected hardware". If we now let API calls float between different Conns, we lose that binding, and we can no longer do any strong credential cloning prevention, as one example.

Reply:
That's fine. If/when we do that, we can do that in the Dial at the beginning before the http2 part starts.

Reply:
Unexported